File: android/IonicPortals/src/main/kotlin/io/ionic/portals/PortalsPlugin.kt
package io.ionic.portals

import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject

@CapacitorPlugin(name = "Portals")
class PortalsPlugin : Plugin() {

    companion object {
        @JvmStatic
        var subscriptions = mutableMapOf<String, MutableMap<Int, (data: Map<String, Any>) -> Unit>>()
        @JvmStatic
        var subscriptionRef = 0
        @JvmStatic
        fun publish(topic: String, data: Any) {
            subscriptions[topic]?.let {
                for((ref, listener) in it) {
                    val result = mutableMapOf<String, Any>(
                        "topic" to topic,
                        "subscriptionRef" to ref,
                        "data" to data
                    )
                    listener(result)
                }
            }
        }

        @JvmStatic
        fun subscribe(topic: String, callback: (data: Map<String, Any>) -> Unit): Int {
            subscriptionRef++
            subscriptions[topic]?.let { subscription ->
                subscription[subscriptionRef] = callback
            } ?: run {
                val subscription = mutableMapOf(subscriptionRef to callback)
                subscriptions[topic] = subscription
            }
            return subscriptionRef
        }

        @JvmStatic
        fun unsubscribe(topic: String, subscriptionRef: Int) {
            subscriptions[topic]?.let { subscription ->
                subscription.remove(subscriptionRef)
            }
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun publishNative(call: PluginCall) {
        val topic = call.getString("topic") ?: run {
            call.reject("topic not provided")
            return
        }
        val data = call.data.get("data");
        if(data is JSONObject) {
            PortalsPlugin.publish(topic, (data as JSONObject).toMap())
        } else {
            PortalsPlugin.publish(topic, data)
        }
        call.resolve()
    }

    @PluginMethod(returnType = PluginMethod.RETURN_CALLBACK)
    fun subscribeNative(call: PluginCall) {
        val topic = call .getString("topic") ?: run {
            call.reject("topic not provided")
            return
        }
        call.setKeepAlive(true)
        val ref = PortalsPlugin.subscribe(topic) { data ->
            call.resolve(JSObject(JSONObject(data).toString()))
        }
        val result = JSObject()
        result.put("topic", topic)
        result.put("subscriptionRef", ref)
        call.resolve(result)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun unsubscribeNative(call: PluginCall) {
        val topic = call .getString("topic") ?: run {
            call.reject("topic not provided")
            return
        }
        val subscriptionRef = call .getInt("subscriptionRef") ?: run {
            call.reject("subscriptionRef not provided")
            return
        }
        PortalsPlugin.unsubscribe(topic, subscriptionRef)
        call.resolve()
    }

}

fun JSONObject.toMap(): Map<String, Any> {
    val map = mutableMapOf<String, Any>();
    this.keys().forEach {
        map[it] = this.get(it)
    }
    return map
}

File: ios/IonicPortals/IonicPortals/PortalsPlugin.swift
import Foundation
import Capacitor

@objc(PortalsPlugin)
public class PortalsPlugin: CAPPlugin {
    
    private static var subscriptions: [String: [Int : (_ result: SubscriptionResult) -> ()]] = [:]
    private static var subscriptionRef = 0
    
    // MARK: Methods used by Cap Plugin

    @objc func publishNative(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic") else {
            call.reject("topic not provided")
            return
        }
        let data = call.getAny("data")
        PortalsPlugin.publish(topic, data!)
        call.resolve()
    }
    
    @objc func subscribeNative(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic") else {
            call.reject("topic not provided")
            return
        }
        call.keepAlive = true
        let ref = PortalsPlugin.subscribe(topic, {result in
            call.resolve(result.toMap())
        })
        call.resolve([
            "topic": topic,
            "subscriptionRef": ref
        ])
    }
    
    @objc func unsubscribeNative(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic") else {
            call.reject("topic not provided")
            return
        }
        guard let subscriptionRef = call.getInt("subscriptionRef") else {
            call.reject("subscriptionRef not provided")
            return
        }
        PortalsPlugin.unsubscribe(topic, subscriptionRef)
        call.resolve()
    }
    
    // MARK: Static methods for use by Swift app
    
    public static func subscribe(_ topic: String, _ callback: @escaping (_ result: SubscriptionResult) -> ()) -> Int {
        PortalsPlugin.subscriptionRef += 1
        if var subscription = PortalsPlugin.subscriptions[topic] {
            subscription[PortalsPlugin.subscriptionRef] = callback
            PortalsPlugin.subscriptions[topic] = subscription
        } else {
            let subscription = [PortalsPlugin.subscriptionRef : callback]
            PortalsPlugin.subscriptions[topic] = subscription
        }
        return PortalsPlugin.subscriptionRef
    }
    
    public static func publish(_ topic: String, _ data: Any) {
        if let subscription = PortalsPlugin.subscriptions[topic] {
            for(ref, listener) in subscription {
                let result = SubscriptionResult(topic: topic, data: data, subscriptionRef: ref)
                listener(result)
            }
        }
    }
    
    public static func unsubscribe(_ topic: String, _ subscriptionRef: Int) {
        if var subscription = PortalsPlugin.subscriptions[topic] {
            subscription.removeValue(forKey: subscriptionRef)
            PortalsPlugin.subscriptions[topic] = subscription
        }
    }
            
}

public struct SubscriptionResult {
    public var topic: String
    public var data: Any
    public var subscriptionRef: Int
    
    func toMap() -> [String: Any] {
        return [
            "topic": self.topic,
            "data": self.data,
            "subscriptionRef": self.subscriptionRef
        ]
    }
}

